java/ql/lib/semmle/code/java/dataflow/internal/DataFlowDispatch.qll

```diff
@@ -40,11 +40,29 @@ private module DispatchImpl {
     else any()
   }
 
-  /** Gets a viable implementation of the target of the given `Call`. */
+  /**
+   * Gets a viable implementation of the target of the given `Call`.
+   * The following heuristic is applied for finding the appropriate callable:
+   * 1. If an exact manual model exists, only dispatch to the summarized callable.
+   * 2. If a (non exact) manual model exists and/or if the source code is available, dispatch to both/either.
+   * 3. Only dispatch to a summarized callable (based on a generated model) if neither of the above apply.
+   */
   DataFlowCallable viableCallable(DataFlowCall c) {
-    result.asCallable() = sourceDispatch(c.asCall())
-    or
-    result.asSummarizedCallable().getACall() = c.asCall()
+    exists(Call call | call = c.asCall() |
+      result.asCallable() = sourceDispatch(call)
+      or
+      not (
+        // Only use summarized callables with generated summaries in case
+        // we are not able to dispatch to a source declaration.
+        // Note that if applyGeneratedModel holds it implies that there doesn't
+        // exist a manual (exact) model.
+        exists(Callable callable | callable = sourceDispatch(call) |
+          callable.fromSource() and not callable.isStub()
+        ) and
+        result.asSummarizedCallable().applyGeneratedModel()
+      ) and
```

Comment:
I'm slightly concerned whether this risks FNs. I.e. if some api point is intended to be overridden then you may have library implementations from which a good model was generated and also source implementations that are very different. I'm not exactly sure how to best iterate on this heuristic, but I think we should require that the static call target is `fromSource` in order to boost our confidence in the source dispatch.

Reply:
To make sure I understand your concern.
Are you worried about the following:
Assume we have a library *L* containing
```
class Base { ... }
class A : Base { ... }
```
and we generate models based on *L*, where the implementation of `A` gives us "good" models for `Base`.

If we furthermore assume that we want to analyse some source code, which imports `L`, but also contains a "bad" implementation of `Base` (that doesn't exhibit the same flow as `A`)
```
class B : Base { ... }
```

Reply:
Yes, exactly. It might very well be that `A` is the default implementation when interacting with the library through `Base` and that `B` is degenerate in some way - e.g. a test stub.
Another condition to potentially take into account is whether the static call target is `B` or `Base`. If the static call target is `B` then it's probably fine to disregard any generated summaries.

Reply:
Fair point, but at least the DCA projects didn't exhibit this problematic pattern.
What do you think about changing the generated models exclusion logic to something like:
```
not (
        // Only use summarized callables with generated summaries in case
        // we are not able to dispatch to a source declaration.
        // Note that if applyGeneratedModel holds it implies that there doesn't
        // exist a manual (exact) model.
        exists(Callable staticTarget | staticTarget = call.getCallee() |
          staticTarget.fromSource() and not staticTarget.isStub()
        ) and
        exists(Callable callable | callable = sourceDispatch(call) |
          callable.fromSource() and not callable.isStub()
        ) and
        result.asSummarizedCallable().applyGeneratedModel()
      )
```
There might also be something special for constructors that needs to be takes into account - but I can try and write some tests for this.

Reply:
Yes, I think that looks much better.